Scheduled Snapshots Feature

koushikcgit opened the pull request:
This PR has all the commits for the VMSS feature (REQ-277)

stephen-turner commented:
I have noted on the xapi pull request, https://github.com/xapi-project/xen-api/pull/2577, that there are terminology problems in the SDK -- places where "scheduled_snapshot" should be "snapshot_schedule". This will affect the function names in this pull request too.

stephen-turner commented on XenModel/XenAPI-Extensions/IVMPolicy.cs:
```diff
@@ -0,0 +1,59 @@
+        IXenConnection Connection { get; }
+        bool is_enabled();
+        bool is_running();
+        bool is_archiving();
```
It would be more idiomatic to make is_enabled, is_running and is_archiving into getters, not functions.

koushikcgit replied:
I kept these as functions as this was what it was earlier in the VMPP.cs (XAPI-Extensions) and in the usage as well. Do we want to change that existing behaviour?

stephen-turner replied:
I hadn't realised this was existing behaviour. But I think we should change it anyway, as it's a small change.

koushikcgit replied:
Indeed it's a small change. Will make the modifications.

stephen-turner commented on XenModel/XenAPI-Extensions/VMSS.cs:
```diff
@@ -0,0 +1,271 @@
+
+        public bool is_running()
+        {
+            return false;
```
Why is is_running always false for VMSS? Is it the case that we have this info for VMPP but not for VMSS for some reason?

koushikcgit replied:
Yes. We do not have is_running in the VMSS object. Hence this is by default 'false'

stephen-turner replied:
OK

stephen-turner commented on XenModel/Messages.resx:
```diff
@@ -11496,6 +11572,14 @@ To learn more about the [XenServer] Performance Reporting capabilities feature o
 To learn more about the [XenServer] Role Based Access Control feature or to start a [XenServer] trial, click the button below.</value>
   </data>
+  <data name="UPSELL_BLURB_VMSS" xml:space="preserve">
+    <value>Upgrade your [XenServer] license to enable VM Snapshot Schedule. VM Snapshot Schedule allows you to create automatic snapshots for your critical VMs.</value>
```
The feature is called "VM Scheduled Snapshots". (This affects five lines below this too, and maybe other places).

koushikcgit replied:
Ah! So there seems to be a disconnect in what we would want to call this feature after all. I recall, from our discussion, that we wanted to have this as `VM Snapshot Schedule` and looks like we want to different now.
I would then prefer to have `VM Scheduled Snapshot` be consistent, in the XC, XAPI, SDK, etc. Does this sound acceptable?

stephen-turner replied:
No, sorry, I don't think that works.

The feature name is "Scheduled Snapshots" (actually, that's better than the "VM Scheduled Snapshots" I used above).
A single VMSS is a "Snapshot Schedule".
A "scheduled snapshot" is a single snapshot, created according to the schedule.

I have made the same comment on xapi-project/xen-api#2577

stephen-turner commented on XenModel/Messages.resx:
```diff
@@ -1874,12 +1874,18 @@ Archive destination: {0}</value>
     <value>Protection policy assigned.</value>
   </data>
+  <data name="ASSIGNED_VMSS_POLICY" xml:space="preserve">
+    <value>Snapshot schedule policy assigned.</value>
```
The phrase "snapshot schedule policy" sounds clumsy to me. I would rather just call it a "snapshot schedule", i.e. without the "policy". (This affects lots of strings).

stephen-turner commented:
We still have a lot of places where we test (typeof(T) == typeof(VMPP)) or (typeof(T) == typeof(VMSS)). The purpose of having the IVMPolicy interface is to do away with all (or almost all) of these. Something like archiving, or quiescing, where the features genuinely differ may be an exception. But for anything where the features are essentially the same, but are using different classes or variable names in the API, the IVMPolicy interface should abstract these differences away leaving the class that's using them clean.

stephen-turner commented:
@koushikcgit, I haven't done a full review of the logic, but above are some higher-level comments on terminology and code design. I think we should address these first and then go into a more detailed review. I realise some of them are in Sharad's code and he's away for three weeks -- is anyone deputising for him?

koushikcgit commented:
@stephen-turner : I think the use of typeof(T) to differentiate VMPP and VMSS is present in areas where the distinction is required either for UI controls, or when different apis need to be invoked in some other objects (like vm, etc.). Let's discuss this in detail to see which ones should be removed.

stephen-turner commented:
@koushikcgit I still think we can get rid of almost all of them. Do we need a GTM to go through some examples?

koushikcgit commented:
@stephen-turner : I've updated the PR. Please have a look

stephen-turner commented on XenAdmin/Wizards/NewPolicyWizard/NewPolicySnapshotFrequencyPage.cs:
```diff
@@ -45,11 +45,12 @@ namespace XenAdmin.Wizards.NewPolicyWizard
     public partial class NewPolicySnapshotFrequencyPage : XenTabPage, IEditPage
     {
-        public NewPolicySnapshotFrequencyPage()
+        public NewPolicySnapshotFrequencyPage(bool isVMSS)
```
This file still needs to have the duplicate code removed.
_copy and _copyVMSS should be declared as IVMGroup. Then all the other instances of if(isVMSS) can be removed. (IVMGroup.frequency will have to return a new enum rather than vmss_frequency or vmpp_backup_frequency).

stephen-turner commented on XenAdmin/Wizards/NewPolicyWizard/NewPolicySnapshotTypePageSpecific.cs:
```diff
@@ -0,0 +1,313 @@
+            return null;
+        }
+
+        private VMPP _clone;
```
This class also needs duplicate code removed.

stephen-turner commented on XenAdmin/Wizards/NewPolicyWizard/NewPolicyWizardSpecific.cs:
```diff
@@ -0,0 +1,303 @@
+            throw new ArgumentException("wrong argument");
+        }
+
+        private static string FormatBackupTypeVMSS(vmss_type backupType)
```
This could be combined with the previous function by defining a new enum for backup type in IVMGroup.

stephen-turner commented on XenAdmin/Wizards/NewPolicyWizard/NewPolicyWizardSpecific.cs:
```diff
@@ -0,0 +1,303 @@
+                    Connection = Pool.Connection
+                };
+
+                var action = new CreateVMSS(vmss, xenTabPageVMsPage.SelectedVMs, xenTabPageFinish.RunNow);
```
I think the CreateVMPP and CreateVMSS blocks can also be combined, via a VMGroup<T>.CreateObjectAction() function. We will pass in the archive data even for VMSS, and it will get ignored later.

stephen-turner commented on XenModel/Actions/VMPP/AssignVMsToPolicyAction.cs:
```diff
@@ -36,57 +36,64 @@
-            : string.Format(Messages.REMOVE_VMS_FROM_POLICY, vmpp.Name))
+        public RemoveVMsFromPolicyAction(IVMPolicy policy, List<XenRef<VM>> selectedVMs)
+            : base(policy.Connection, selectedVMs.Count == 1 ?
+            (typeof(T) == typeof(VMPP) ? string.Format(Messages.REMOVE_VM_FROM_POLICY, policy.Connection.Resolve(selectedVMs[0]), policy.Name) :
```
I'm happy to have the typeof(T) == typeof(VMPP) here, but let's move it inside the string.Format for conciseness.

stephen-turner commented on XenModel/Actions/VMPP/ChangePolicyEnabledAction.cs:
```diff
@@ -39,24 +39,39 @@
+            bool value = !_policy.is_enabled;
+            if (typeof(T) == typeof(VMPP))
+            {
+                Description = value ? string.Format(Messages.ENABLING_VMPP, _policy.Name) : string.Format(Messages.DISABLING_VMPP, _policy.Name);
```
Again in this area, let's move the typeof(T) == typeof(VMPP) inside the string.Format.

stephen-turner commented on XenModel/Actions/VMPP/RunPolicyNowAction.cs:
```diff
@@ -36,20 +36,21 @@
         {
-            Result = VMPP.protect_now(Session, _vmpp.opaque_ref);
-            Description = string.Format(Messages.RUN_POLICY_STARTED, _vmpp.Name);
+            var _vmpp = _policy as VMPP;
```
_vmpp seems to be an unused variable. This won't compile in Release build.

stephen-turner commented on XenModel/Actions/VMSS/CreateVMSS.cs:
```diff
@@ -0,0 +1,75 @@
+﻿/* Copyright (c) Citrix Systems Inc. 
```
I think the CreateVMSS action and the GetVMSSAlertsAction should not be needed: we should template CreateVMPP and GetVMPPAlertsAction instead.

stephen-turner commented on XenModel/FriendlyNames.resx:
```diff
@@ -1214,6 +1403,24 @@
     <value>Unable to log in to XAPI session.</value>
   </data>
+  <data name="Message.body-vmss_license_error" xml:space="preserve">
+    <value>The XenServer license you are using does not allow you to use the Scheduled Snapshot feature.</value>
```
Scheduled Snapshots feature (in the plural).

stephen-turner commented on XenModel/FriendlyNames.resx:
```diff
@@ -546,18 +582,45 @@
   <data name="Label-host.edition-advanced" xml:space="preserve">
     <value>[XenServer product] Advanced Edition</value>
   </data>
+  <data name="Label-host.edition-desktop" xml:space="preserve">
```
There are a lot of new strings in this file, and they don't appear to be all new or recent features. Where did they come from? We need to understand this before we can merge them.

stephen-turner commented on XenModel/Messages.resx:
```diff
@@ -2409,6 +2427,16 @@ Do you want to assign these VMs to the policy '{0}' instead?</value>
 Do you want to assign it to the policy '{2}' instead?</value>
   </data>
+  <data name="CHANGE_VMS_SNAPSHOT_SCHEDULE_WARNING" xml:space="preserve">
+    <value>Some of the VMs you selected are already assigned to a different snapshot schedule policy.
```
Delete "policy"

stephen-turner commented on XenModel/Messages.resx:
```diff
@@ -1874,12 +1874,18 @@ Archive destination: {0}</value>
     <value>Protection policy assigned.</value>
   </data>
+  <data name="ASSIGNED_VMSS_POLICY" xml:space="preserve">
+    <value>Snapshot Schedule assigned.</value>
```
There's a lot of inconsistency whether we write Snapshot Schedule with or without upper case letters. I recommend standardising on lower case.

stephen-turner commented on XenModel/Messages.resx:
```diff
@@ -8430,6 +8485,12 @@ It is strongly recommended that you Cancel and apply the latest version of the p
+    <value>Use this wizard to define a VM protection policy that will ensure your critical VMs are protected via automatic scheduled snapshot jobs.</value>
+  </data>
+  <data name="NEW_VMSS_PAGE_TEXT" xml:space="preserve">
+    <value>Use this wizard to define a VM Snapshot Schedule that will create automatic schedule snapshot jobs for your VMs.</value>
```
"automatic scheduled" (not "schedule").

stephen-turner commented on XenModel/Messages.resx:
```diff
@@ -3253,6 +3284,9 @@ Please reconnect the host and try again.</value>
     <value>VM protection policy '{0}' created.</value>
   </data>
+  <data name="CREATED_VMSS" xml:space="preserve">
+    <value>VM Snapshot Schedule '{0}' created.</value>
```
I think we should always refer to just "Snapshot schedule" not "VM snapshot schedule"

stephen-turner commented on XenModel/XenAPI-Extensions/VMSS.cs:
```diff
@@ -0,0 +1,305 @@
+
+        private static DateTime GetDailyDate(DateTime time, int min, int hour)
+        {
+            var nextDateTime = new DateTime(time.Year, time.Month, time.Day, hour, min, 0);
```
In this file, any function which is static (or could be static like TryGetKey), we can just call the VMPP version directly, we don't need to make another copy of the code.

stephen-turner commented:
Thank you for your updates, @koushikcgit. This looks much better now. I hope you agree that this approach makes the code much cleaner.

I think we're close now. There are still a couple more places where we could avoid code duplication, which I've mentioned above, along with a couple of other things.

We haven't tried out the code yet. I'll try and find time to do that tomorrow.

stephen-turner commented:
What is the plan for merging this? I guess we need to wait until the xapi changes are in trunk?

GaborApatiNagy commented:
It looks like github could not merge this PR  because there are conflicts,
could you rebase/merge master into your branch as well, please? Thanks.

cheng-z commented on XenAdmin/Dialogs/VMProtectionRecovery/VMProtectionPoliciesDialog.cs:
```diff
@@ -123,8 +88,8 @@ private class PolicyRow : DataGridViewRow
             private DataGridViewTextAndImageCell _lastResult = new DataGridViewTextAndImageCell();
-            public readonly VMPP VMPP;
-            public PolicyRow(VMPP policy)
+            public readonly IVMPolicy _Policy;
```
_Policy -> _policy will be better I think.

koushikcgit commented:
Stephen, this can be merged to trunk without the xapi changes as the feature depends on the `restrict_vmss` license param.

stephen-turner commented:
@koushikcgit The code looks good now, thank you.

I did find a couple of bugs while I was trying it out on your test server:

1. Select a VM that already has a snapshot schedule; "assign to snapshot schedule" from the VM menu and choose another schedule; get error "Object has been deleted.VMSS:"
2. The New VMSS Wizard has "policy" in lots of places. First, even before you get into the wizard, the menu item is called "New policy"; four times on the first tab of the wizard; on the Snapshot Type tab (where it says "Protection Policy").

stephen-turner commented:
For the merge strategy, it does leak the name "snapshot schedule" in the UI, even if it's greyed out, so I would prefer to wait to merge until the feature is ready end-to-end. We can either keep this PR open, or merge it onto a branch.

stephen-turner commented on XenModel/XenAPI-Extensions/VMPP.cs:
```diff
@@ -166,6 +167,95 @@ public override string Description
+
+        public policy_backup_type policy_type 
+        {
+            get { return (policy_backup_type)backup_type; }
```
This type of code is not really safe. It only happens to work because the policy_backup_type enum is in the same order as the vmpp_backup_type enum, with the additional items not present in vmpp_backup_type at the end. But you shouldn't make such assumptions about enums in case they get extended in a different way later.

I realise it's longer code, but it would be more idiomatic not to convert between different types of enums and to use a switch statement to enumerate through the backup types and return the right policy_backup type for each one (and vice versa in the setter).

koushikcgit commented:
@stephen-turner : Thanks for the review. I've updated the PR and things should now converge.
On the bugs you uncovered:
1. I've already seen this earlier and have filed a CA for this. Needs to be looked into from xapi side.
2. I've fixed these texts in the update to the PR

On the merge strategy:
1. Yes, I agree that it will leave the name "snapshot schedule" in the UI. Out of the suggestions you have, what is that you prefer?

stephen-turner commented:
Great, thanks. I'll look at it later today.

Re merging, I think my preference is to create a new branch and merge it onto there, and then merge it into master when the xapi work is done.

stephen-turner commented:
@koushikcgit, there was just one more little thing I noticed. I'm not sure we need the "Quiescing Supported" column in the New VMSS Wizard. Feel free to disagree if you think I'm wrong, but it seems to me to give it too much prominence. It's already covered on the next page by virtue of the option being greyed out or not.

stephen-turner commented:
I have created a "VMSS" branch which is currently identical to master. Could you raise the PR against that branch instead? Then we can keep it on there until we're ready to merge it into master. Thank you.

koushikcgit commented:
Creating new PR for merging into VMSS branch - https://github.com/xenserver/xenadmin/pull/850
Closing this one and referenced this into the new PR for the comments
